File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        st, en, answer = 0, 0, 0
        sub = set()
        while en < len(s):
            if s[en] in sub:
                sub.remove(s[st])
                st += 1
            else:
                sub.add(s[en])
                en += 1
                answer = max(answer, len(sub))
        return answer

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.kt
class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        var start = 0
        var maxLen = 0
        val set = mutableSetOf<Char>()
        for (end in s.indices) {
            while (set.contains(s[end])) {
                set.remove(s[start])
                start++
            }
            set.add(s[end])
            val len = end - start + 1
            if (maxLen < len) {
                maxLen = len
            }
        }
        return maxLen
    }
}